### DISCUSSION TITLE
How do I render the RepeatedType "invalid_message" on both fields?

### DISCUSSION THREAD

**Discussion by AZeroEight** [symfony/symfony#48014] [Q&A] [answered]
According to the documentation on [Symfony.com](https://symfony.com/doc/current/reference/forms/types/repeated.html) you can create a **RepeatedType** field to enforce two matching passwords:
```php
$builder->add('password', RepeatedType::class, [
    'type' => PasswordType::class,
    'invalid_message' => 'The password fields must match.',
    'options' => ['attr' => ['class' => 'password-field']],
    'required' => true,
    'first_options'  => ['label' => 'Password'],
    'second_options' => ['label' => 'Repeat Password'],
]);
```
...this is working well and I can also add constraints to the field:
```php
...
'constraints' => [
  new Length([
    'min' => 6,
    'minMessage' => "needs to be longer"
    'max' => 4096
  ]),
],
...
```
...the constraints are rendered on both fields, but the _invalid_message_ only on the first one - how do I make it appear on both fields?

**Comment by MatTheCat** [comment]
I think the best way is to pass the constraint in `options`:

```php
$builder->add('password', RepeatedType::class, [
    'type' => PasswordType::class,
    'invalid_message' => 'The password fields must match.',
    'options' => [
        'attr' => ['class' => 'password-field'],
        'constraints' => new Length([
            'min' => 6,
            'minMessage' => "needs to be longer"
            'max' => 4096
        ])
    ],
    'required' => true,
    'first_options'  => ['label' => 'Password'],
    'second_options' => ['label' => 'Repeat Password'],
]);
```

**Reply by AZeroEight** [reply]
That part is working fine - the constraints you add as "options" work as expected and there is no need to pass them on as first_options and second_options. The part that is not rendering on both fields is the message that they don't match - passed on as **invalid_message**. That message only shows up on the first field.

**Reply by MatTheCat** [reply]
Damn sorry I read your question too fast. I guess you would have to override the `repeated_row` block in your form template but not sure how yet.

**Reply by MatTheCat** [reply]
Okay I got it working by extending the `RepeatedType`:

```php
<?php

namespace App\Form\Extension;

use Symfony\Component\Form\AbstractTypeExtension;
use Symfony\Component\Form\Extension\Core\Type\RepeatedType;
use Symfony\Component\Form\FormInterface;
use Symfony\Component\Form\FormView;
use Symfony\Component\OptionsResolver\OptionsResolver;

class RepeatedTypeExtension extends AbstractTypeExtension
{
    public function finishView(FormView $view, FormInterface $form, array $options)
    {
        foreach ($view->children as $child) {
            $child->vars['errors'] = $view->vars['errors'];
        }
    }

    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefault('error_mapping', []);
    }

    public static function getExtendedTypes(): iterable
    {
        return [RepeatedType::class];
    }
}
```

It overwrites `RepeatedTypeValidatorExtension`’s `error_mapping` to keep errors on the main field, and copies them to each child. As the template does not display the main field’s errors (because they are supposed to be mapped to the first child), we’re all set!

**Reply by AZeroEight** [reply]
So far it only removed the error message on the fields but you put me into the right direction and I will give it a try this evening. I did not know about the powerful possibility to extend the types and need to do further reading. I will share my solution once I got it working - if nobody else is faster :smile:

**Reply by MatTheCat** [reply]
Hm it’s weird I got it working on my computer :thinking: 

Be aware that if you use this extension, constraints must be defined on the `RepeatedType`, not its children (this should not cause any issue as you would want the same).

**Reply by AZeroEight** [reply]
They are defined on the RepeatedType - just as seen in the example on top. But if I dump the variables, the value of ```$view->vars['errors']``` has an array 'errors' which is empty while the one at the first child is not:
```
Symfony\Component\Form\FormView {#1218 ▼
  +vars: array:33 [▼
    ...
    "errors" => Symfony\Component\Form\FormErrorIterator {#1215 ▼
      -form: Symfony\Component\Form\Form {#815 ▶}
      -errors: []
    ...
  +children: array:2 [▼
    "first" => Symfony\Component\Form\FormView {#1214 ▼
      +vars: array:33 [▼
        ...
        "errors" => Symfony\Component\Form\FormErrorIterator {#1210 ▼
          -form: Symfony\Component\Form\Form {#818 ▶}
          -errors: array:1 [▼
            0 => Symfony\Component\Form\FormError {#1217 ▶}
```

...so in this case I would need to copy the error from the first to the second field, but I only want to do that if I encounter the issue that the two passwords don't match. The other constraints (not long enough or empty) should be specific to the field causing the issue.

The following will copy the message from the first to the second child and since the other constraints only matter if both fields are the same, we are probably getting away with this. To make the styles match as well I needed two lines:

```php
$view->children['second']->vars['errors'] = $view->children['first']->vars['errors'];
$view->children['second']->vars['valid']    = $view->children['first']->vars['valid'];
```

I can't tell if this solution is messy or not due to lag of experience using Symfony but if this solution is fine with you, I would like to ask you to either change your post or incorporate the changes so that i can mark your solution as the solution to this issue.

**Reply by MatTheCat** [reply]
Indeed your implementation is simpler than mine :ok_hand: and FWIW it works on my computer too (as long as the constraint is in `options.constraints`).

So I guess you would use it like this:

```php
<?php

namespace App\Form\Extension;

use Symfony\Component\Form\AbstractTypeExtension;
use Symfony\Component\Form\Extension\Core\Type\RepeatedType;
use Symfony\Component\Form\FormInterface;
use Symfony\Component\Form\FormView;

class RepeatedTypeDuplicateErrorsExtension extends AbstractTypeExtension
{
    public function finishView(FormView $view, FormInterface $form, array $options)
    {
        $view->children['second']->vars['errors'] = $view->children['first']->vars['errors'];
    }

    public static function getExtendedTypes(): iterable
    {
        return [RepeatedType::class];
    }
}
```

**Reply by AZeroEight** [reply]
Hey  @MatTheCat,

I'd like to close this - could you please post your code with the changes discussed so you get the credit once I mark it as the correct solution?

Thank you very much for your help.

**Reply by MatTheCat** [reply]
Well you're the one who found the best solution so go ahead and add your own answer, I promise I won't be mad 😁

**Comment by AZeroEight** [comment, marked as answer]
Create a file ```Form/Extension/RepeatedTypeExtension.php```
```php
<?php

namespace App\Form\Extension;

use Symfony\Component\Form\AbstractTypeExtension;
use Symfony\Component\Form\Extension\Core\Type\RepeatedType;
use Symfony\Component\Form\FormInterface;
use Symfony\Component\Form\FormView;

class RepeatedTypeExtension extends AbstractTypeExtension
{
    public function finishView(FormView $view, FormInterface $form, array $options)
    {
        // make sure errors are shown on both fields:
        $view->children['second']->vars['errors'] = $view->children['first']->vars['errors'];
        $view->children['second']->vars['valid'] = $view->children['first']->vars['valid'];
    }

    public static function getExtendedTypes(): iterable
    {
        return [RepeatedType::class];
    }
}
```

Thanks to @MatTheCat who pointed me into the right direction!
